### PR TITLE
Adding repair step for sub share

### PR DESCRIPTION
**Pull request by sharidas**
Adding repair step for sub share.
When share table due to some magic gets
duplicate entries for sub share ( except the id or primary
key ), then this repair step comes for rescue.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description

Remove the duplicate entries from the `share` table which can cause exception described in the issue.

## Related Issue

https://github.com/owncloud/core/issues/27990

## Motivation and Context

For some magical reason, if the share table gets populated with duplicate entries then we would have to remove the duplicate entry. Keeping the row with minimum id. Else exception would be thrown as mentioned in the issue.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by sharidas** [comment]
Need to write unit test. But so far with the scenario mentioned in the issue, when I run the `maintenance:repair` command the output is as follows:
```
sujith@sujith-Inspiron-5567 ~/test/owncloud2 $ ./occ maintenance:repair 
Cannot load Xdebug - it was already loaded
ownCloud is in maintenance mode - no app have been loaded

 - Repair mime types
 - Detect file cache entries with path that does not match parent-child relationships
 - Generate ETags for file where no ETag is present.
     - ETags have been fixed for 0 files/folders.
 - Clean tags and favorites
     - 0 tags of deleted users have been removed.
     - 0 tags for delete files have been removed.
     - 0 tag entries for deleted tags have been removed.
     - 0 tags with no entries have been removed.
 - Drop old database tables
 28/28 [============================] 100%
 - Drop old background jobs
 - Remove getetag entries in properties table
     - Removed 0 unneeded "{DAV:}getetag" entries from properties table.
 - Repair outdated OCS IDs
 - Repair invalid shares
 - Remove old share propagation app entries
 - Move user avatars outside the homes to the new location
 - Remove shares of a users root folder
 - Repair unmerged shares
 - Disable extra themes
 - Repair sub shares
     - Removed 1 shares where duplicate rows where found
sujith@sujith-Inspiron-5567 ~/test/owncloud2 $ ./occ maintenance:repair 
Cannot load Xdebug - it was already loaded
ownCloud is in maintenance mode - no app have been loaded

 - Repair mime types
 - Detect file cache entries with path that does not match parent-child relationships
 - Generate ETags for file where no ETag is present.
     - ETags have been fixed for 0 files/folders.
 - Clean tags and favorites
     - 0 tags of deleted users have been removed.
     - 0 tags for delete files have been removed.
     - 0 tag entries for deleted tags have been removed.
     - 0 tags with no entries have been removed.
 - Drop old database tables
 28/28 [============================] 100%
 - Drop old background jobs
 - Remove getetag entries in properties table
     - Removed 0 unneeded "{DAV:}getetag" entries from properties table.
 - Repair outdated OCS IDs
 - Repair invalid shares
 - Remove old share propagation app entries
 - Move user avatars outside the homes to the new location
 - Remove shares of a users root folder
 - Repair unmerged shares
 - Disable extra themes
 - Repair sub shares
sujith@sujith-Inspiron-5567 ~/test/owncloud2 $
```

**Comment by sharidas** [comment]
Created the tests for this repair PR.

File: lib/private/Repair/RepairSubShares.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		$builder
+			->delete('share')
+			->where($builder->expr()->eq('id', $builder->createParameter('shareId')));
+
+		$this->deleteShareId = $builder;
+	}
+
+	public function run(IOutput $output) {
+		$deletedEntries = 0;
+		$this->removeRowsWithDuplicateDataExceptId();
```
Comment (PVince81) [inline lib/private/Repair/RepairSubShares.php]: this function's name is lying. It is not removing rows with duplicate data except id. It's only setting some attributes

Comment (sharidas) [reply lib/private/Repair/RepairSubShares.php]: @PVince81 Yes its making the query statement with attributes. Will get it corrected. It was there in my mind sometime back and then got slipped :( Thanks.

File: tests/lib/Repair/RepairSubSharesTest.php [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+			\OC::$server->getGroupManager()->get($groupName.$groupCount)->addUser($user);
+
+			//Create a group share
+			$qb->insert('share')
+				->values([
+					'share_type' => $qb->expr()->literal('2'),
+					'share_with' => $qb->expr()->literal($userName.$groupCount),
+					'uid_owner' => $qb->expr()->literal('admin'),
+					'uid_initiator' => $qb->expr()->literal('admin'),
+					'parent' => $qb->expr()->literal(1),
```
Comment (PVince81) [inline tests/lib/Repair/RepairSubSharesTest.php]: I suggest you also vary the parent value considering that we use it on the `group by`

File: tests/lib/Repair/RepairSubSharesTest.php [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+					'item_type' => $qb->expr()->literal('folder'),
+					'item_source' => $qb->expr()->literal(24),
+					'file_source' => $qb->expr()->literal(24),
+					'file_target' => $qb->expr()->literal($folderName.$groupCount),
+					'permissions' => $qb->expr()->literal(31),
+					'stime' => $qb->expr()->literal($time),
+				])
+				->execute();
+
+			if (($i%3) === 0) {
```
Comment (PVince81) [inline tests/lib/Repair/RepairSubSharesTest.php]: a bit more fresh air appreciated

**Review by PVince81 — COMMENTED** [review]
Looks fine, see comment

**Comment by PVince81** [comment]
any update ?

**Review by PVince81 — APPROVED** [review]
:+1:

@tom @butonic any comments on the query from https://github.com/owncloud/core/pull/30186/files#diff-c7d3960633d9f6fe9eb517281d947301R61 regarding performance ?

**Comment by PVince81** [comment]
@sharidas please backport

**Comment by sharidas** [comment]
Backport PR: https://github.com/owncloud/core/pull/30695

**Comment by phil-davis** [comment]
Backport PR has been done and merged.

**Pull request merged**